src/Analyzer.Reports/ReportsHelper.cs

```diff
@@ -0,0 +1,121 @@
+            IEnumerable<string> filesToSkip)
+        {
+            return GetResultsByFile(evaluations, filesToSkip, out int _);
+        }
+
+        /// <summary>
+        /// Get all distinct results in an evaluation organized by source location file
+        /// </summary>
+        /// <param name="evaluations">Evaluations to get results for</param>
+        /// <param name="filesToSkip">Files to not include results from</param>
+        /// <param name="passedEvaluations">Out parameter that gives number of passed evaluations</param>
+        /// <returns></returns>
+        public static Dictionary<string, List<(IEvaluation, IList<IResult>)>> GetResultsByFile(
+            IEnumerable<Types.IEvaluation> evaluations,
+            IEnumerable<string> filesToSkip,
+            out int passedEvaluations)
+        {
+            var resultsByFile = new Dictionary<string, List<(IEvaluation, IList<IResult>)>>();
+            passedEvaluations = 0;
+
+            foreach (var evaluation in evaluations)
+            {
+                if (!evaluation.Passed)
+                {
+                    (var actualFile, var failedResults) = GetResultsByFileInternal(evaluation);
+
+                    // a file's results may have already been output if analyzing directory
+                    if (filesToSkip.Contains(actualFile))
+                    {
+                        continue;
+                    }
+
+                    if (!resultsByFile.ContainsKey(actualFile))
+                    {
+                        resultsByFile[actualFile] = new List<(IEvaluation, IList<IResult>)>();
+                    }
+
+                    // skip any evaluations with duplicate results (i.e. two source locations from other templates refer to same result)
+                    if (!resultsByFile[actualFile].Any(resultsTuple => evaluation.RuleId.Equals(resultsTuple.Item1.RuleId)
+                        && Enumerable.SequenceEqual(resultsTuple.Item2, failedResults)))
+                    {
+                        resultsByFile[actualFile].Add((evaluation, failedResults));
+                    }
+                }
+                else
+                {
+                    passedEvaluations++;
+                }
+            }
+
+            foreach (var file in resultsByFile.Keys)
+            {
+                resultsByFile[file] = resultsByFile[file].Distinct().ToList();
+            }
+
+            return resultsByFile;
+        }
+
+        /// <summary>
+        /// Gets all failed results in an evaluation, sorts, and returns rule ID
+        /// </summary>
+        /// <param name="evaluation"></param>
+        /// <returns></returns>
+        /// <exception cref="Exception"></exception>
+        public static (string, IList<IResult>) GetResultsByFileInternal(Types.IEvaluation evaluation)
+        {
+            // get all distinct failed results in evaluation
+            var failedResults = GetFailedResultsAsList(evaluation).Distinct().ToList();
```

Comment:
This should probably be implemented in the Evaluation: `evaluation.FailedResults()`, which can return distinct failed results.  Using a `HashSet<Result>` will also get rid of the need to call `Distinct()`.

Reply:
Since IEvaluation is an interface, I did this using an extension method.

Also, the tests depend on GetFailedResults for getting the unduped line numbers (the spiritual successor to GetFailedLines), so I need to do the deduping outside of that method.

I got tripped up on this for a while because I thought I was getting different results from List.Distinct() and a HashSet when it was just the test dependency I forgot about. Would have been caught right away if I did the extension method first. 😅